DREAM/Optimization/tsgGradientDescent.hpp

```diff
@@ -0,0 +1,204 @@
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_GRADIENT_DESCENT_HPP
+#define __TASMANIAN_GRADIENT_DESCENT_HPP
+
+#include "tsgOptimizationUtils.hpp"
+
+/*!
+ * \internal
+ * \file tsgGradientDescent.hpp
+ * \brief Gradient descent state and algorithm.
+ * \author Weiwei Kong & Miroslav Stoyanov
+ * \ingroup TasmanianOptimization
+ *
+ * Definition of the gradient descent state class and the particle swarm algorithm.
+ * \endinternal
+ */
+
+namespace TasOptimization {
+
+/*!
+ * \ingroup OptimizationState
+ * \brief Stores the information about a gradient descent run.
+ *
+ * \par class GradientDescentState
+ * An object of this class represents the components of a gradient descent run and all key parts of the swarm can be read through
+ * member methods. Methods are also provided to modify a subset of its parts.
+ *
+ * \par Constructors and Copy/Move assignment
+ * All constructors require information about an initial candidate point and initial stepsize. A method is available to retrieve
+ * the number of dimensions. No default constructor is available for this class, but copy/move constructors and assignment operator=
+ * overloads are available. The number of dimensions \b cannot be modified.
+ * - GradientDescentState()
+ * - getNumDimensions()
+ *
+ * \par Set and Get Candidate and Stepsize
+ * Throughout the run of gradient descent, the candidate point and stepsize may change. Methods are available to retrieve these
+ * data and set them manually.
+ * - getCandidate(), setCandidate()
+ * - getStepsize(), setStepsize()
+ *
+ * \par Gradient Descent Algorithm
+ * The gradient descent optimization algorithm can be used on a gradient descent algorithm with the goal of minimizing a particular
+ * objective function. The current candidate point, with respect to the given objective function, can be obtained from the state
+ * using the getCandidate() method.
+ * - GradientDescent()
+ *
+ * \par
+ * More information about the gradient descent algorithm can be found in the following paper:
+ *
+ * \par
+ * > Nesterov, Y. (2013). Gradient methods for minimizing composite functions. <em>Mathematical programming</em>, 140(1), 125-161.
+ *
+ */
+class GradientDescentState {
+  public:
+    //! \brief The default constructor is NOT allowed.
+    GradientDescentState() = delete;
+    //! \brief Constructor for a gradient descent state with the initial candidate and stepsize.
+    GradientDescentState(const std::vector<double> &x0, const double lambda0);
+    //! \brief Copy constructor.
+    GradientDescentState(const GradientDescentState &source) = default;
+    //! \brief Move constructor.
+    GradientDescentState(GradientDescentState &&source) = default;
+
+    //! \brief Move assignment.
+    GradientDescentState& operator=(GradientDescentState &&source) = default;
```

Comment:
How about copy-assignment, we should allow that one too.

Reply:
Sure. I have added copy assignment to the header.